schema/mysql/schema.sql

```diff
@@ -0,0 +1,296 @@
+
+    CONSTRAINT key_incident_contact_contact UNIQUE (incident_id, contact_id),
+    CONSTRAINT key_incident_contact_contactgroup UNIQUE (incident_id, contactgroup_id),
+    CONSTRAINT key_incident_contact_schedule UNIQUE (incident_id, schedule_id),
+    CONSTRAINT nonnulls_incident_recipients_check CHECK (if(contact_id IS NULL, 0, 1) + if(contactgroup_id IS NULL, 0, 1) + if(schedule_id IS NULL, 0, 1) = 1)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+
+CREATE TABLE incident_rule (
+    incident_id bigint NOT NULL REFERENCES incident(id),
+    rule_id bigint NOT NULL REFERENCES rule(id),
+
+    CONSTRAINT pk_incident_rule PRIMARY KEY (incident_id, rule_id)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+
+CREATE TABLE incident_rule_escalation_state (
+    incident_id bigint NOT NULL REFERENCES incident(id),
+    rule_escalation_id bigint NOT NULL REFERENCES rule_escalation(id),
+    triggered_at bigint NOT NULL,
+
+    CONSTRAINT pk_incident_rule_escalation_state PRIMARY KEY (incident_id, rule_escalation_id)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+
+CREATE TABLE incident_history (
+    id bigint NOT NULL AUTO_INCREMENT,
+    incident_id bigint NOT NULL REFERENCES incident(id),
+    rule_escalation_id bigint REFERENCES rule_escalation(id),
+    event_id bigint REFERENCES event(id),
+    contact_id bigint REFERENCES contact(id),
+    contactgroup_id bigint REFERENCES contactgroup(id),
+    schedule_id bigint REFERENCES schedule(id),
+    rule_id bigint REFERENCES rule(id),
+    channel_id bigint REFERENCES channel(id),
+    time bigint NOT NULL,
+    message text,
+    -- Order to be honored for events with identical millisecond timestamps.
+    type enum('opened', 'incident_severity_changed', 'rule_matched', 'escalation_triggered', 'recipient_role_changed', 'closed', 'notified') NOT NULL,
+    new_severity enum('ok', 'debug', 'info', 'notice', 'warning', 'err', 'crit', 'alert', 'emerg'),
+    old_severity enum('ok', 'debug', 'info', 'notice', 'warning', 'err', 'crit', 'alert', 'emerg'),
+    new_recipient_role enum('recipient', 'subscriber', 'manager'),
+    old_recipient_role enum('recipient', 'subscriber', 'manager'),
+    notification_state enum('pending', 'sent', 'failed'),
+    sent_at bigint,
+
+    CONSTRAINT pk_incident_history PRIMARY KEY (id),
+    FOREIGN KEY (incident_id, rule_escalation_id) REFERENCES incident_rule_escalation_state(incident_id, rule_escalation_id)
+) ENGINE=InnoDB DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_bin;
+
+CREATE INDEX idx_incident_history_time_type ON incident_history(time, type) COMMENT 'Incident History ordered by time/type';
+
+CREATE TABLE browser_session (
+    php_session_id varchar(256) NOT NULL,
+    username varchar(254) NOT NULL,
+    user_agent varchar(255) NOT NULL,
+    authenticated_at bigint NOT NULL,
+
+    CONSTRAINT pk_browser_session PRIMARY KEY (php_session_id, username, user_agent)
```

Comment:
@ncosta-ic Don't say you REALLY need such a combined PK?

Reply:
Yes, I do. I initially used a small [jooat hash](https://en.wikipedia.org/wiki/Jenkins_hash_function#one_at_a_time) for the calculated device id (see https://github.com/Icinga/icinga-notifications/pull/136#discussion_r1407742077), but other contributors wanted me to change it to include the full useragent string (which is truncated to the [first 4096 chars by the PHP daemon](https://github.com/Icinga/icinga-notifications-web/blob/9f9517701ff8b83dd98387a8a903f4b99aee9bb0/library/Notifications/ProvidedHook/SessionStorage.php#L42)).

Reply:
How much of the user agent is stored isn't relevant here to me. It's rather the fact that it's part of the constraint. If the session id is part of it, the user agent is redundant. If we really add it, it would mean that session hijacking is supported.

Reply:
@nilmerg Valid point. I think we included the user agent to narrow down the possibility of a duplicate session identifier, as we are not able to detect a session cleanup peformed by PHP's session routine reliably.
Thus one identifier could still exist in our database for up to 24 hours after it got cleaned up by PHP.
However, we later introduced an additional check on the session creation, which [cleans up an old entry in the database](https://github.com/Icinga/icinga-notifications-web/blob/a6063d1959435398b08b46acdb586cb95be75857/library/Notifications/ProvidedHook/SessionStorage.php#L47-L68) if it happens to have the same php session id as the current connection.

I think we might be able to completely get rid of the user agent in the constraint.

Reply:
Then let's remove it.